firmware-support/bittide-sys/src/time/self_test.rs

```diff
@@ -0,0 +1,150 @@
+    // Construct a list of tests with their names.
+    let tests = [
+        (now_not_null as fn(Clock) -> bool, "now_not_null"),
+        (freq_not_null as fn(Clock) -> bool, "freq_not_null"),
+        (wait_1ms as fn(Clock) -> bool, "wait_1ms"),
+        (skip_next_ms as fn(Clock) -> bool, "skip_next_ms"),
+        (duration_hour as fn(Clock) -> bool, "duration_hour"),
+        (duration_minute as fn(Clock) -> bool, "duration_minute"),
+        (duration_second as fn(Clock) -> bool, "duration_second"),
+        (
+            duration_millisecond as fn(Clock) -> bool,
+            "duration_millisecond",
+        ),
+        (
+            duration_microsecond as fn(Clock) -> bool,
+            "duration_microsecond",
+        ),
+        (instant_minute as fn(Clock) -> bool, "instant_minute"),
+        (instant_second as fn(Clock) -> bool, "instant_second"),
+        (
+            instant_millisecond as fn(Clock) -> bool,
+            "instant_millisecond",
+        ),
+        (
+            instant_microsecond as fn(Clock) -> bool,
+            "instant_microsecond",
+        ),
+    ];
+    // Run the tests and collect the results.
+    let mut results = [("", false); 13];
+    for (i, test) in tests.iter().enumerate() {
+        let (f, name) = test;
+        let result = f(clock.clone());
+        results[i] = (name, result);
+    }
+    results
+}
+
+//// Obtain the value of the counter, check if it's not 0.
+pub fn now_not_null(clock: Clock) -> bool {
+    let frequency = clock.get_frequency();
+    let now = clock.elapsed();
+    return now != Instant::from_cycles(0, frequency);
+}
+
+//// Read the frequency value, check if it's not 0.
+pub fn freq_not_null(clock: Clock) -> bool {
+    let freq = clock.get_frequency();
+    freq != 0
+}
+
+/// Read the current time in milliseconds, wait a ms and read again.
+/// The new time should differ less than 100 us from the expected target.
+pub fn wait_1ms(clock: Clock) -> bool {
+    let wait_time = Duration::from_millis(1);
+    let time0 = clock.elapsed();
+    clock.wait(wait_time);
+    let time1 = clock.elapsed();
+    let expected = time0 + wait_time;
+    time1 - expected < Duration::from_micros(100)
+}
+
+/// Read the current time in milliseconds and create a target of the current time in ms + 2.
+/// Wait until we reach the target and obtain the new time. The time should differ no more than
+/// 100 ms from the target.
+pub fn skip_next_ms(clock: Clock) -> bool {
+    let frequency = clock.get_frequency();
+    let time0 = clock.elapsed();
+    let target = Instant::from_millis(time0.to_millis() + 2, frequency);
+    clock.wait_until(target);
+    let time1 = clock.elapsed();
+    time1 - target < Duration::from_micros(100)
+}
+
+/// Create a Duration of 1 hour, convert back to one hour and check if it's the same.
+pub fn duration_hour(_: Clock) -> bool {
+    let duration = Duration::from_hours(1);
+    let hours = duration.to_hours();
+    hours == 1 && hours != 0
+}
```

Comment:
This function (and all following ones) don't actually need a `Clock` instance, at most they need a frequency. This means that all those functions be run as unit tests on the host too. For the ones needing a frequency, we could just give it a constant value and also run them on the host as unit tests. Anything only using `Duration` or `Instant` doesn't need to be tested on the hardware but can be run on the host.

Reply:
Yes I'm aware, I added it as an argument anyway to keep the type signature the same, allowing me to iterate over the functions in a uniform way.

Would you prefer I separate those iterators into functions with clocks and functions without clocks?